Feature branch: Allow attaching files from Blended Data Repository

amanda-anderson opened the pull request:
Feature branch: Allow attaching files from Blended Data.

All code has been previously reviewed before being merged into this branch.

amanda-anderson committed:
Replace existing solution with new endpoints for phase attachments

amanda-anderson committed:
Fix types

amanda-anderson committed:
Remove deadcode

amanda-anderson committed:
Remove deadcode

amanda-anderson committed:
Fix duplicate messages error

amanda-anderson committed:
Merge in master

amanda-anderson committed:
WIP Add select from data repo dropdown

Commit:
Translations updated by CI (extract-intl)

amanda-anderson committed:
Add select dropdown if data repo enabled

amanda-anderson committed:
Merge branch 'TAN-5078-allow-selecting-files-from-repo-phases' of github.com:CitizenLabDotCo/citizenlab into TAN-5078-allow-selecting-files-from-repo-phases

Commit:
Translations updated by CI (extract-intl)

amanda-anderson committed:
Re-enable screen reader support for file attachments

amanda-anderson committed:
Merge branch 'TAN-5078-allow-selecting-files-from-repo-phases' of github.com:CitizenLabDotCo/citizenlab into TAN-5078-allow-selecting-files-from-repo-phases

amanda-anderson committed:
Fix spelling

amanda-anderson committed:
Generalize useSyncFiles

amanda-anderson committed:
Code cleanup

amanda-anderson committed:
Code cleanup

amanda-anderson committed:
Code cleanup

amanda-anderson committed:
Use new file attachment endpoints for project attachments

amanda-anderson committed:
Remove deadcode

amanda-anderson committed:
Code cleanup

amanda-anderson committed:
Remove deadcode

amanda-anderson committed:
Code cleanup

amanda-anderson committed:
Code cleanup

amanda-anderson committed:
WIP - Allow attaching from data repo for events

amanda-anderson committed:
Remove files from event form

amanda-anderson committed:
Code cleanup

amanda-anderson committed:
Merge branch 'master' into TAN-5078-allow-selecting-files-from-repo-phases

amanda-anderson committed:
Merge branch 'TAN-5078-allow-selecting-files-from-repo-phases' into TAN-5443-allow-selecting-files-from-repo-projects

amanda-anderson committed:
Merge branch 'TAN-5443-allow-selecting-files-from-repo-projects' into TAN-5442-allow-selecting-files-from-repo-events

amanda-anderson committed:
Remove deadcode

amanda-anderson committed:
Add missing prop

amanda-anderson committed:
Improve loading state and code cleanup

amanda-anderson committed:
Only show existing file selector if there are file options

jinjagit committed:
[TAN-5442] naive approach - not considering project

jinjagit committed:
[TAN-5442] Consider project for update and delete policy

jinjagit committed:
[TAN-5442] Consider project for create policy

jinjagit committed:
[TAN-5442] Fix policy spec

amanda-anderson committed:
Remove console logs

jinjagit committed:
[TAN-5442] Actually use variable as intended

jinjagit committed:
[TAN-5442] Use blank not empty

jinjagit committed:
[TAN-5442] Create files_project when create attachment, if needed

jinjagit committed:
[TAN-5442] Wrap before_save & save in transaction

jinjagit committed:
[TAN-5442] Remove transaction and move files_project creation into controller action

jinjagit committed:
[TAN-5442] Any elevated role can create files_file

jinjagit committed:
[TAN-5442] Remove redundant method param

jinjagit committed:
[TAN-5442] Fix specs

jinjagit committed:
[TAN-5442] Simplify policy method

jinjagit committed:
Merge branch 'TAN-5442-allow-selecting-files-from-repo-events' into TAN-5442-file-policy-for-pms

jinjagit committed:
[TAN-5442] Explicit check for roles

jinjagit committed:
[TAN-5442] Do not remove private section

amanda-anderson committed:
Add flag to tell if files have been changed

amanda-anderson committed:
Merge pull request #12088 from CitizenLabDotCo/TAN-5442-allow-selecting-files-from-repo-events

[TAN-5442] Allow selecting files from repository on Events

amanda-anderson committed:
Merge pull request #12072 from CitizenLabDotCo/TAN-5443-allow-selecting-files-from-repo-projects

[TAN-5443] Use new File Attachment endpoints for Project attachments

amanda-anderson committed:
Code cleanup

amanda-anderson committed:
Merge in master

amanda-anderson committed:
Merge pull request #12035 from CitizenLabDotCo/TAN-5078-allow-selecting-files-from-repo-phases

[TAN-5078] Use new File Attachment endpoints for Phase attachments

amanda-anderson committed:
Merge remote-tracking branch 'origin/TAN-5442-file-policy-for-pms' into feature-branch-blended-data-file-attachments-v2

amanda-anderson committed:
Code cleanup & invalidate file attachments after file deletion

amanda-anderson committed:
WIP fix E2E tests

amanda-anderson committed:
Fix e2e test

amanda-anderson committed:
Fix circular update warning

amanda-anderson committed:
Fix e2e test

amanda-anderson committed:
Merge branch 'TAN-5442-file-policy-for-pms' of github.com:CitizenLabDotCo/citizenlab into TAN-5442-file-policy-for-pms

amanda-anderson committed:
Stabilize test

amanda-anderson committed:
Merge pull request #12105 from CitizenLabDotCo/TAN-5442-file-policy-for-pms

[TAN-5442] Files::File policy for pms

amanda-anderson committed:
Merge branch 'master' into feature-branch-blended-data-file-attachments-v2

amanda-anderson committed:
Implement BE so normal users and logged out visitors can see file widget attachment if permitted

amanda-anderson committed:
Add files to included array and add file attachments to files relationships

amanda-anderson committed:
Add link in sense-making to Files Tab

Commit:
Translations updated by CI (extract-intl)

amanda-anderson committed:
Fix be spec

amanda-anderson committed:
Merge pull request #12125 from CitizenLabDotCo/TAN-5505-add-files-included-array

[Tan-5505] Add Files to file_attachment included array.

amanda-anderson committed:
Merge branch 'master' into feature-branch-blended-data-file-attachments-v2

amanda-anderson committed:
Merge branch 'feature-branch-blended-data-file-attachments-v2' of github.com:CitizenLabDotCo/citizenlab into feature-branch-blended-data-file-attachments-v2

amanda-anderson committed:
Fix import

amanda-anderson committed:
Merge branch 'feature-branch-blended-data-file-attachments-v2' into TAN-5322-file-attachment-be

amanda-anderson committed:
Fix failing tests by re-introducing the moderator check

amanda-anderson committed:
Fix BE lint issue

amanda-anderson committed:
Add better warning when deleting files

Commit:
Translations updated by CI (extract-intl)

amanda-anderson committed:
code cleanup

amanda-anderson committed:
Merge branch 'TAN-5117-v4-warning-file-deletion' of github.com:CitizenLabDotCo/citizenlab into TAN-5117-v4-warning-file-deletion

amanda-anderson committed:
Merge pull request #12123 from CitizenLabDotCo/TAN-5322-file-attachment-be

[TAN-5322] File Attachment Widget - BE

amanda-anderson committed:
Copy Adrien's fix for the attachment deletion issue

amanda-anderson committed:
Merge pull request #12131 from CitizenLabDotCo/blended-data-fix-deletion-issue

Blended data: Fix for file attachment deletion deleting file

amanda-anderson committed:
Stabilize e2e test

amanda-anderson committed:
Merge pull request #12132 from CitizenLabDotCo/TAN-5117-v4-warning-file-deletion

Tan 5117 v4 warning file deletion

adessy commented on back/app/policies/files/file_attachment_policy.rb:
```diff
@@ -29,6 +29,8 @@ def create?
       when 'Phase', 'Event'
         active_moderator?(record.attachable.project)
+      when 'ContentBuilder::Layout'
+        active_moderator?(Project.find(record.attachable.project_id))
```
This should not be necessary. I'll try to revert it to the original version.

adessy commented on back/app/policies/files/file_policy.rb:
```diff
@@ -3,6 +3,8 @@
   class FilePolicy < ApplicationPolicy
     class Scope < ApplicationPolicy::Scope
+      # TODO: Need to update this method with the same logic as the show method,
+      # as signed out users now have access to some files (those attached to public content).
```
It was intentional. The goal was to use the `GET .../file_attachments` endpoints to list the files in the front-office. Attachments should in principle contain all the necessary information to display the files (position/ordering, name, size, url). At least, this was the intention.

adessy commented on back/app/policies/files/file_policy.rb:
```diff
@@ -14,38 +16,42 @@ def resolve
+        # Can the user moderate at least one of the associated projects
+        return UserRoleService.new.moderatable_projects(user, record.projects).exists?
+      end
+
+      # Can the user see whatever the file is attached to
+      record.attachments.any? do |attachment|
+        policy_for(attachment.attachable).show?
+      end
     end
```
Same here, this restriction was intentional.

adessy commented on back/app/policies/files/file_policy.rb:
```diff
@@ -14,38 +16,42 @@ def resolve
-      return false if projects.empty?
-
-      UserRoleService.new.moderatable_projects(user, projects).count == projects.count
+      admin? || user.project_or_folder_moderator? # Any elevated role can create
```
I don't think this is correct. Unless I'm mistaken, this would allow moderators to create files in projects they don't manage.

adessy commented on back/app/policies/files/file_policy.rb:
```diff
@@ -14,38 +16,42 @@ def resolve
 
-      # Allow update if the user moderates at least one of the associated projects.
-      UserRoleService.new.moderatable_projects(user, record.projects).exists?
+      moderates_all_projects?
```
Not sure what the issue was here. Currently, files are always associated with a single project. So, in practice, those two implementations are equivalent.

adessy commented on back/app/controllers/web_api/v1/files/files_controller.rb:
```diff
@@ -2,6 +2,7 @@
 
 class WebApi::V1::Files::FilesController < ApplicationController
   include CarrierwaveErrorDetailsTransformation
+  skip_before_action :authenticate_user, only: %i[index show]
```
Same same :) The goal was to use only attachments in the front office, not files.

adessy commented on back/app/controllers/web_api/v1/files/file_attachments_controller.rb:
```diff
@@ -30,6 +32,18 @@ def show
   def create
     file_attachment = Files::FileAttachment.new(create_params)
+
+    # We permit creation of files_files records before creation of an associated project (i.e. in project create/edit form).
```
Not quite sure I understand why we allow the creation of files outside the scope of a project (in other words, files without `files_projects`).

adessy reviewed (commented):
Hey @amanda-anderson  I  went through the back-end code and I think a few things might need some fixing. See my comments below.

There seems to be a mix of misunderstandings and a few remaining issues in my implementation.

The main misunderstanding (though in the end, we might have to go this way if we don't want to revert those changes) is about the permissions of the new files endpoints. Those permissions were quite strict because they were not meant to be used by the front office, except for one thing: the creation of idea files. For everything else, the intention was that the front office would only use the file attachment endpoints.

Now, the bug in my implementation is that the permissions for the create file endpoint are also too strict. But this didn’t really matter or cause issues since we didn’t refactor the front office to use the new endpoints for file creation.

There are a few other things I’m not sure I fully understand:
- In which cases do we want to create the project-file association ex-post, and why?
- Why do file upload and file attachment need to happen in two steps in some cases? You touched on this during the meeting yesterday morning, but I didn’t fully get it.
- Why didn’t we include this [commit](https://github.com/CitizenLabDotCo/citizenlab/commit/33f7f860a1d3054d722b238a562855c13512671c) to remove idea files from the 360 Input tab in the back office? Is it not needed anymore?

adessy committed:
Add file details attributes to FileAttachmentSerializer

adessy committed:
Revert 57c1f9de80b138ce9e373a4c752f1327d2ac581c

Normal users or visitors do not need to be able to show files, only file attachments, as they contain all the file details relevant for the front-office.

adessy committed:
Refactor `user_moderates_all_projects?` in FilePolicy

adessy committed:
Revert e7932bb3b590cfd504a0b98eae41b9ded6555a7d and 1988a2ce7dc261e07da255654a61f6bec91868fd

Those cases should be covered by:
```ruby
return false unless policy_for(record.attachable).update?
```

adessy committed:
Refactor file and file attachment policies

amanda-anderson committed:
Use file attachment in FO for file widget

amanda-anderson committed:
WIP file widget back office

amanda-anderson committed:
Cleanup file attachments properly for file attachment widget deletion

amanda-anderson committed:
Remove unnecessary comment

amanda-anderson committed:
Code cleanup

adessy committed:
Prevent files from being added to multiple projects

We may reconsider this in the future, but for the sake of simplicity, we decided not to allow it for now.

amanda-anderson committed:
Improve typing for useFileAttachmentById hook

amanda-anderson committed:
Move content builder delete cleanup function and make it more generalized/extendable

amanda-anderson committed:
Merge pull request #12156 from CitizenLabDotCo/fe-changes-finalization-1

Use file_attachments endpoint in FO, and improve cleanup when file widgets deleted

amanda-anderson committed:
Merge branch 'master' into feature-branch-blended-data-file-attachments-v2

amanda-anderson committed:
Small bug fixes

amanda-anderson committed:
Do not shrink paperclip icon in file display if long file name

adessy committed:
Fix the on-the-fly linking of global files to a project when attached to a project's resource

adessy committed:
Merge branch 'feature-branch-blended-data-file-attachments-v2' into blended-data-repo/finalization-1

# Conflicts:
#	back/db/structure.sql

amanda-anderson committed:
Merge pull request #12147 from CitizenLabDotCo/blended-data-repo/finalization-1

[Input360] Finalization 1

amanda-anderson committed:
Merge branch 'master' into feature-branch-blended-data-file-attachments-v2

amanda-anderson commented:
Ran the E2E tests - just the two that are already failing on master failed, so we are all good :+1:
